### ISSUE TITLE
[PyTorch/XLA] Improve the XLA PR landing process

### ISSUE THREAD

**Issue by JackCaoG** [pytorch/pytorch#82033]
### 🚀 The feature, motivation and pitch

# Current situation
Currently to merge a XLA breaking pr, we follow these procedures
1. PyTorch/XLA merge the fix PR, which will break the PyTorch/XLA head CI
2. PyTorch PR rebased and updated the [xla_hash](https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt)
3. PyTorch merge the xla break pr
4. PyTorch/XLA rerun the head CI and it should be green again

# User expericncen flaw
This process has help a lot for pytorch CI stability. However it creates quite a lot of burden on the PyTorch/XLA maintainer side because
1. PyTorch/XLA CI will be broken when fix pr merged and won't be fix until corresponding PyTorch PR merged. 
2. Merging a pytorch pr is a lot more effort than merging a pytorch/xla PR. In step 2 pytorch pr needs to be rebased hence will rerun all test. Waiting for all tests to finish will take hours and might fail for various reason. In this whole time PyTorch/XLA head will be broken
3. using `merge -f` will help this merging delay but people are generally concern about this command because it is somewhat dangerous for such a complcaited codebase like pytorch.
4. PyTorch/XLA maintainer needs to closely monitor the upstream pr until it is merged.

# Proposal 
I am proposing a new merge process hopefully can address above issue
1. PyTorch/XLA rebased the "fix pr"
2. PyTorch pr rebased and pin to the PyTorch/XLA pr's branch in [xla.txt](https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt)
3. PyTorch pr started a merge job and can wait until all test green(xla test should pass too)
4. PyTorch/XLA merge the fix pr (need to turn off auto-delete branch after merge)
5. PyTorch bot will update the [xla.txt](https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt) to the new PyTorch/XLA master hash which will also have the fix

The advantage of this approach ares
1. PyTorch and PyTorch/XLA pr will be both green all the time
2. PyTorch/XLA maintainer only need to merge the fix pr the same day(before pytorch bot started). This requires much less coordination work between two sides.

# Work that needs to be done
PyTorch CI need to understand  [xla.txt](https://github.com/pytorch/pytorch/blob/master/.github/ci_commit_pins/xla.txt) can contain both branch name and commit.

FYI @suo @bdhirsh @wconstab

**Comment by huydhn** [comment]
In the new process proposal, would putting the commit hash of PyTorch/XLA "fix pr" branch into xla.txt instead of the branch name work? If it does, the new workflow can be tried out right away without much development work I think.

* PyTorch CI only [checkout](https://github.com/pytorch/pytorch/blob/f50a248a5eacb9a9aa475a9e610486aea136e4f5/.jenkins/pytorch/common_utils.sh#L119) the commit hash in xla.txt. This should work in the same way for all XLA branches, including the master and the "fix pr" one.
* The same "limitation" applies in which PyTorch/XLA maintainers only need to merge the fix pr the same day before PyTorch bot starts. On the other hand, the bot could be improved a bit there to **not** overwrite the commit hash (with the one from master branch) if the timestamp of the latest commit on XLA master branch is older (than the current one in xla.txt). This ensures that the correct commit will be used until it's merged into XLA master (even when the process takes more than a day for any reasons)

**Comment by janeyx99** [comment]
also cc’ing @clee2000 regarding the hash logic

**Comment by JackCaoG** [comment]
@huydhn oh right.. I guess checkout can handle both branch and commit. There is a landing in progress right now(pytorch/xla pr merged, waiting for pytorch one to be merged). I will start to experiment the new process starting next PR. 

Regarding your second point I think it sounds good to me.

Btw should there be a doc somewhere regarding the merge process? This way people can check the latest landing process themselves. Happy to put that doc in pytorch/xla repo if you guys think it is not suitable for pytorch main repo. 

FYI @wonjoolee95

**Comment by huydhn** [comment]
> Btw should there be a doc somewhere regarding the merge process? This way people can check the latest landing process themselves. Happy to put that doc in pytorch/xla repo if you guys think it is not suitable for pytorch main repo.

I couldn't find any wiki on this. @janeyx99 Do you have one in mind? Otherwise, let's create one :)

**Comment by clee2000** [comment]
afaik the closest thing we have to a wiki about this is https://docs.google.com/document/d/1jRT-on-v8bcTCHD8aGFH7IuCDR7JL37GlP6Z2EYgmyk/edit, but this is more of a doc about why the hash pinning is done rather than a guide on how to merge 

i can get started on the changes for the hash pinning script regarding not updating a pinned commit if the pinned commit is newer

**Comment by ZainRizvi** [comment]
> * The same "limitation" applies in which PyTorch/XLA maintainers only need to merge the fix pr the same day before PyTorch bot starts. On the other hand, the bot could be improved a bit there to **not** overwrite the commit hash (with the one from master branch) if the timestamp of the latest commit on XLA master branch is older (than the current one in xla.txt). This ensures that the correct commit will be used until it's merged into XLA master (even in the case in which the process takes more than a day for any reasons)

Would this work even if a new commit gets merged to pytorch/xla's master branch? Pytorch/xla gets ~3-7 commits per day.

Timeline I'm thinking of:
1. Breaking change gets committed to pytorch/pytorch with sha A
2. pytorch/pytorch pins xla at sha A
3. Someone else merges a new commit to pytorch/xla with sha B
4. The PR from sha A isn't merged before midnight
5. At midnight, the bot pins sha B to pytorcy/pytorch since it has a newer timestamp

**Comment by huydhn** [comment]
> Would this work even if a new commit gets merged to pytorch/xla's master branch? Pytorch/xla gets ~3-7 commits per day

Yeah, you're spot on. The nightly run CI has no way to know if it should keep the sha A from the not yet merged "fix pr" branch, or the latest (unrelated) commit from XLA master (if any) that is merged before sha A is merged. So it's not a perfect solution. On the other hand, having a timestamp check is still better than nothing. It's easy to implement and covers the basic case I guess :)
